### PR TITLE
bin: Naive implementation for explain config or secrets

### PR DESCRIPTION
**Pull request by aarnq**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

I hacked this together so I might as well share it.

```console
$ ./bin/ck8s explain config
Compliant Kubernetes Apps settings

This describes the structure of the configuration for both the service
cluster and the workload cluster, but keep in mind that each configuration
file will contain different settings.

type: object

properties:
  - global

defaults: null

examples: null
```

#### Information to reviewers

This will most likely become slower as the schema becomes larger as the dereferencing is a bit naive, but it isn't really possible to just do it across the whole file in one go.
Maybe internal anchoring and aliasing might solve it, but I believe that one cannot simply explode recursive references :thinking: 

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [x] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests
- Config checks:
  - [ ] The schema was updated

### CODE REVIEW

File: bin/explain.bash [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  array)
+    yq4 '{"properties": .items.properties | keys}' <<< "${data}"
+    echo
+    ;;
+  object)
+    yq4 '{"properties": .properties | keys}' <<< "${data}"
+    echo
+    ;;
+  esac
+  yq4 '{"defaults": .default}' <<< "${data}"
```
Comment (Zash) [inline bin/explain.bash:99]: Worth suppressing `defaults: null`, or leave that as motivation for adding that to the schema?

I'm less sure everything will have defaults, especially higher level maps where the defaults may be contained in each property.

File: bin/explain.bash [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  echo
+  yq4 '{"examples": .examples}' <<< "${data}"
```
Comment (Zash) [inline bin/explain.bash:101]: Worth suppressing `defaults: null`, or leave that as motivation for adding that to the schema?

File: bin/explain.bash [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  data="$(recvdereference "$1")"
+
+  if [[ -n "${2:-}" ]]; then
+    local -a path
+    readarray -t path <<< "$(yq4 'split "." | .[]' <<< "$2")"
+
+    for key in "${path[@]}"; do
+      case "$(yq4 ".type" <<< "${data}")" in
+      array)
+        data="$(yq4 ".items.properties.${key}" <<< "${data}")" ;;
```
Comment (Zash) [inline bin/explain.bash:62]: Assuming arrays always containing objects?

File: bin/explain.bash [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    ;;
+  object)
+    yq4 '{"properties": .properties | keys}' <<< "${data}"
+    echo
+    ;;
+  esac
+  yq4 '{"defaults": .default}' <<< "${data}"
+  echo
+  yq4 '{"examples": .examples}' <<< "${data}"
+}
```
Comment (Zash) [inline bin/explain.bash:102]: Wishlist: Showing things like `enum` if they're included would be nice.

**Review by Zash — APPROVED** [review]
Looks great!

Have some non-blocking questions and comments.

File: bin/explain.bash [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  log_error "usage: ck8s explain <config|secrets> [key.to.parameter]"
+  exit 1
+}
+
+dereference() {
+  # shellcheck disable=SC2016
+  yq4 '. as $root | with(
+    .. | select(key == "$ref");
+    . = "$root." + (
+      sub("#/", "") | split "/" | with(.[]; . = "\"" + . + "\"" ) | join "."
```
[the remark this answers is not in the document]
Comment (aarnq) [reply bin/explain.bash:27]: Yeah, we could probably do a search and replace to decode `~1` to `/` after the join if these become used within the key name. But for now lets note it down as a limitation.

File: bin/explain.bash (hunk 1 above)
Comment (aarnq) [reply bin/explain.bash:99]: Yes, for types `object` and `array` without properties/items.

File: bin/explain.bash (hunk 2 above)
Comment (aarnq) [reply bin/explain.bash:101]: Yes, defaults will be more important than examples, so I'm fine with suppressing them if they don't exist.

File: bin/explain.bash (hunk 3 above)
Comment (aarnq) [reply bin/explain.bash:62]: :sweat_smile:

File: bin/explain.bash (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zash) [reply bin/explain.bash:27]: Also `"` may need special treatment? Will this work? Probably fine to leave it to a future iteration.

```suggestion
      sub("#/", "") | split "/" | with(.[]; . = ( . | to_json) ) | join "."
```

[the remark this answers is not in the document]
Comment (aarnq) [reply bin/explain.bash:27]: Can we like _not_ accept quotes in keys? :sweat_smile:
Though if that works it is much better than the string addition.

[the remark this answers is not in the document]
Comment (Zash) [reply bin/explain.bash:27]: Something we can enforce in that future pre-commit schema validation hook you mentioned? That'd be sensible.

File: bin/explain.bash (hunk 2 above)
Comment (Zash) [reply bin/explain.bash:101]:
```suggestion
  if yq4 --exit-status '.examples' <<< "{data}" >/dev/null 2>&1; then
    yq4 '{"examples": .examples}' <<< "${data}"
  fi
```

File: bin/explain.bash (hunk 4 above)
Comment (Zash) [reply bin/explain.bash:102]: (but leaving that to a future improvement PR is fine)

Comment (aarnq) [reply bin/explain.bash:159]: Implemented, also did special handling for `objects of *` with additional parameters  and `arrays of *` with item examples.

File: bin/explain.bash [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  done
+
+  echo "${data}"
+}
+
+explain() {
+  local data
+  data="$(recvdereference "$1")"
+
+  if [[ -n "${2:-}" ]]; then
```
Comment (Xartos) [inline bin/explain.bash:54]: Suggestion: Move this positional variable to a named one

```suggestion
explain() {
  explainKey="${2}"
  ...
  if [[ -n "${explainKey:-}" ]]; then
```

File: bin/explain.bash [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      object)
+        if [[ "${key}" == "additionalProperties" ]]; then
+          data="$(yq4 ".additionalProperties" <<< "${data}")"
+        else
+          data="$(yq4 ".properties.${key}" <<< "${data}")"
+        fi
+        ;;
+      *)
+        log_error "unable to navigate to ${key} on path to $2 found unnavigable type: $(yq4 ".type" <<< "${data}")"
+        exit
```
Comment (Xartos) [inline bin/explain.bash:72]: Exit with fail code?
```suggestion
        exit 1
```

**Review by Xartos — APPROVED** [review]

**Pull request merged**